app_dart/test/service/content_aware_hash_service_test.dart

```diff
@@ -249,6 +249,84 @@ void main() {
       expect(result, MergeQueueHashStatus.error);
     });
   });
+
+  group('markArtifactsAsComplete', () {
+    setUp(() {
+      firestoreService.putDocument(
+        ContentAwareHashBuilds(
+          createdOn: DateTime.now(),
+          contentHash: '1' * 40,
+          commitSha: 'a' * 40,
+          buildStatus: BuildStatus.inProgress,
+          waitingShas: ['b' * 40, 'c' * 40, 'd' * 40],
+        ),
+      );
+    });
+
+    test('does what it says on the tin', () async {
+      final shas = await cahs.completeArtifacts(
+        commitSha: 'a' * 40,
+        successful: true,
+      );
+      expect(shas, ['b' * 40, 'c' * 40, 'd' * 40]);
+      expect(
+        firestoreService,
+        existsInStorage(ContentAwareHashBuilds.metadata, [
+          isContentAwareHashBuilds
+              .hasCommitSha('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa')
+              .hasContentHash('1' * 40)
+              .hasStatus(BuildStatus.success),
+        ]),
+      );
+    });
+
+    test('does not complete already completed', () async {
+      firestoreService.putDocument(
+        ContentAwareHashBuilds(
+          createdOn: DateTime.now(),
+          contentHash: '1' * 40,
+          commitSha: 'a' * 40,
+          buildStatus: BuildStatus.success,
+          waitingShas: ['b' * 40, 'c' * 40, 'd' * 40],
+        ),
+      );
+      final shas = await cahs.completeArtifacts(
+        commitSha: 'a' * 40,
+        successful: true,
+        maxAttempts: 1,
+      );
+      expect(shas, isEmpty);
+      expect(firestoreService.rollbacks, hasLength(1));
```

Comment:
Just a thought, does not have to be in this PR:

What if instead we threw an error if a transaction was ever _not_ comitted and not rolled back?

Reply:
So instead of `try/catch{rollback}`? It does throw if the retries fail 5 times in a row since those would not get committed or rolled back.

Or - are you asking that fake firestore has a teardown handler that sees if there are any transactions that have not been committed rolled back?

Reply:
The latter.